src/goto-analyzer/static_simplifier.cpp

```diff
@@ -0,0 +1,303 @@
+
+    remove_skip(goto_functions);  // Remove all of the new skips
+    goto_functions.update();
+  }
+
+  status() << "Writing goto binary" << eom;
+  return write_goto_binary(out, ns.get_symbol_table(), goto_functions);
+}
+
+
+/*******************************************************************\
+
+Function: static_simplifiert<analyzerT>::simplify_guard
+
+  Inputs: An iterator pointing to an instruction.
+
+ Outputs: 1 if simplified, 0 if not.
+
+ Purpose: Simplifies the instruction's guard using the information in
+          the abstract domain.
+
+\*******************************************************************/
+
+template<class analyzerT>
+unsigned static_simplifiert<analyzerT>::simplify_guard(
+  goto_programt::instructionst::iterator &i_it)
+{
+  exprt simplified=domain[i_it].domain_simplify(i_it->guard, ns);
+  unsigned return_value=(simplified==i_it->guard)?0:1;
+  i_it->guard=simplified;
+  return return_value;
+}
+
+/*******************************************************************\
+
+Function: static_simplifiert<analyzerT>::simplify_program
+
+  Inputs: None.
+
+ Outputs: None.
+
+ Purpose: Simplifies the program using the information in the abstract
+          domain.
+
+\*******************************************************************/
+
+template<class analyzerT>
+void static_simplifiert<analyzerT>::simplify_program()
+{
+  struct counterst
+  {
+    unsigned asserts;
+    unsigned assumes;
+    unsigned gotos;
+    unsigned assigns;
+    unsigned function_calls;
+  };
+
+  counterst simplified={0, 0, 0, 0, 0};
+  counterst unmodified={0, 0, 0, 0, 0};
+
+  Forall_goto_functions(f_it, goto_functions)
+  {
+    Forall_goto_program_instructions(i_it, f_it->second.body)
+    {
+      if(i_it->is_assert())
+      {
+        unsigned result=simplify_guard(i_it);
+        simplified.asserts+=result;
+        unmodified.asserts+=(1-result);
+      }
+      else if(i_it->is_assume())
+      {
+        unsigned result=simplify_guard(i_it);
+        simplified.assumes+=result;
+        unmodified.assumes+=(1-result);
+      }
+      else if(i_it->is_goto())
+      {
+        unsigned result=simplify_guard(i_it);
+        simplified.gotos+=result;
+        unmodified.gotos+=(1-result);
+      }
+      else if(i_it->is_assign())
+      {
+        code_assignt assign(to_code_assign(i_it->code));
+
+        /*
+        Simplification needs to be aware of which side of the
+        expression it is handling as:
+        <i=0, j=1>  i=j
+        should simplify to i=1, not to 0=1.
+        */
+
+        exprt simp_lhs=domain[i_it].domain_simplify(assign.lhs(), ns, true);
+        exprt simp_rhs=domain[i_it].domain_simplify(assign.rhs(), ns, false);
+
+        unsigned result=(simp_lhs==assign.lhs() &&
+                         simp_rhs==assign.rhs())?0:1;
+
+        simplified.assigns+=result;
+        unmodified.assigns+=(1-result);
+
+        assign.lhs()=simp_lhs;
+        assign.rhs()=simp_rhs;
+        i_it->code=assign;
```

Comment:
To be honest, there are too many redundant expression copies in this branch. How about making "assign" a non-const reference, then that last line could go away. (And the first one would not be a copy.)